File: lib/rtsupport.ml
open Ast

let sourcepath = "./bin/rt-support.c"

let ast =
  [Ast.Fundecl({
    typ = Ast.TypI;
    fname = "getint";
    formals = [];
    body = None;
  });
  Ast.Fundecl({
    typ = Ast.TypV;
    fname = "print";
    formals = [(Ast.TypI, "num")];
    body = None;
  });
  Ast.Fundecl({
    typ = Ast.TypV;
    fname = "printchar";
    formals = [(Ast.TypC, "ch")];
    body = None;
  });
  Ast.Fundecl({
    typ = Ast.TypV;
    fname = "printbool";
    formals = [(Ast.TypB, "b")];
    body = None;
  });]

let iter iterfn = List.iter iterfn ast

File: bin/rt-support.c
/*
  Run-time support library of MicroC
*/
#include <stdio.h>
#include <stdlib.h>

int getint() {
    char buffer[32];
    if (fgets(buffer, 32, stdin) == NULL)
      return 0;
    return atoi(buffer);
}

void print(int num) {
  printf("%d\n", num);
}

void printchar(char ch) {
  printf("%c", ch);
}

void printbool(int b) {
  if (b) {
    printf("true\n");
  } else {
    printf("false\n");
  }
}